packages/core/src/routes/session/passwordless.ts

```diff
@@ -135,208 +136,94 @@ export default function passwordlessRoutes<T extends AnonymousRouter>(
+    await assignInteractionResults(ctx, provider, { login: { accountId: id } });
 
-      assertThat(
-        await hasUserWithPhone(phone),
-        new RequestError({ code: 'user.phone_not_exists', status: 422 })
-      );
+    return next();
+  });
 
-      await verifyPasscode(jti, PasscodeType.SignIn, code, { phone });
-      const { id } = await findUserByPhone(phone);
-      ctx.log(type, { userId: id });
+  router.post(`${signInRoute}/email`, async (ctx, next) => {
+    const verificationStorage = await getVerificationStorageFromInteraction(ctx, provider);
 
-      await updateUserById(id, { lastSignInAt: Date.now() });
-      await assignInteractionResults(ctx, provider, { login: { accountId: id } }, true);
+    const type = getPasswordlessRelatedLogType(PasscodeType.SignIn, 'email');
+    ctx.log(type, verificationStorage);
 
-      return next();
-    }
-  );
+    const { email } = emailSignInSessionGuard(verificationStorage);
 
-  router.post(
-    `${signInRoute}/email/send-passcode`,
-    koaGuard({ body: object({ email: string().regex(emailRegEx) }) }),
-    async (ctx, next) => {
-      const { jti } = await provider.interactionDetails(ctx.req, ctx.res);
-      const { email } = ctx.guard.body;
-      const type = 'SignInEmailSendPasscode';
-      ctx.log(type, { email });
+    assertThat(
+      await hasUserWithEmail(email),
+      new RequestError({ code: 'user.email_not_exists', status: 422 })
+    );
+    const { id } = await findUserByEmail(email);
+    ctx.log(type, { userId: id });
 
-      assertThat(
-        await hasUserWithEmail(email),
-        new RequestError({ code: 'user.email_not_exists', status: 422 })
-      );
+    await updateUserById(id, { lastSignInAt: Date.now() });
+    await assignInteractionResults(ctx, provider, { login: { accountId: id } });
 
-      const passcode = await createPasscode(jti, PasscodeType.SignIn, { email });
-      const { dbEntry } = await sendPasscode(passcode);
-      ctx.log(type, { connectorId: dbEntry.id });
-      ctx.status = 204;
+    return next();
+  });
 
-      return next();
-    }
-  );
+  router.post(`${registerRoute}/sms`, async (ctx, next) => {
+    const verificationStorage = await getVerificationStorageFromInteraction(ctx, provider);
 
-  router.post(
-    `${signInRoute}/email/verify-passcode`,
-    koaGuard({ body: object({ email: string().regex(emailRegEx), code: string() }) }),
-    async (ctx, next) => {
-      const { jti } = await provider.interactionDetails(ctx.req, ctx.res);
-      const { email, code } = ctx.guard.body;
-      const type = 'SignInEmail';
-      ctx.log(type, { email, code });
+    const type = getPasswordlessRelatedLogType(PasscodeType.Register, 'sms');
+    ctx.log(type, verificationStorage);
 
-      assertThat(
-        await hasUserWithEmail(email),
-        new RequestError({ code: 'user.email_not_exists', status: 422 })
-      );
+    const { phone } = smsRegisterSessionGuard(verificationStorage);
```

Comment:
nit: This guard is never been reused, I'll suggest writing in-line.

Reply:
updated.